---
title: "Final Assignment Experimental Design and Data Management (2223-MSB1005)"
author: "Ram√≥n Reszat"
date: "2022-12-18"
output: html_document
---

This notebook creates an annotated differential gene expression dataset that is suitable for downstream biological analysis of cardiomyopathy by pathway or network analysis. The original dataset contains the transcriptomes of left ventricular free-wall tissues from non-failing heart donors and patients with different types of cardiomyopathy extracted during cardiac surgery.

![](background_cardiomyopathy.png)

```{r background, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
| Systolic Dysfunction         | Normal       | Diastolic Dysfunction  |
|:---------------|:---------------------:|:-----------------|
| Dilated cardiomyopathy (DCM)       | Not-failing heart (NF) | Hypertrophic cardiomyopathy (HCM) |
"
cat(tabl)
```		

```{r libraries, include=FALSE}
require(reshape2) # melt function
require(stargazer) # summary statistics
library(ggplot2) # plotting
library(grid) # showing data frames as tables
library(gridExtra) # formatting tables in html
library(pcaMethods) # Bioconductor PCA
require(ltm) # biserial correlation
require(rcompanion) # Cramer's V
library(limma) # Differential Gene Expression Analysis
library(EnhancedVolcano) # visualizing differential gene expression
library(biomaRt) # download Ensemble gene annotations
library(pheatmap) # relative gene expression heatmap
```

## Data Import
First, load the [MAGnet dataset](https://www.med.upenn.edu/magnet) and its annotations on patients and genes. This section is a short exploratory data analysis on the meta-information provided with the MAGnet dataset.
```{r dataset, warning=FALSE}
setwd('MAGNET_GX_Assignment_2022')

# transcriptome of each sample from the NCBI MAGnet dataset
gxData <- read.delim('MAGNET_GeneExpressionData_CPM_19112020.txt', header=TRUE, sep='\t', row.names=1)

# participant information for the tissue samples in the dataset
sampleInfo <- read.csv('MAGNET_SampleData_18112022.csv',header=TRUE, stringsAsFactor=TRUE)

# exon length of each gene for normalization of CPM to TPM
exonLength <- read.delim('MAGNET_exonLengths.txt', header=TRUE, sep='\t', row.names=1)
```

```{r results, warning=FALSE}
# create a directory for the results in the current wd
dir.create(file.path(getwd(), 'MAGNET_GX_Results_2022'))
```

We would like to have an overview of the **participant characteristics** for the dataset in a publication ready table. We can calculate *descriptive statistics* for the continuous variables in the sample information table. Then, we can group the data by etiology to get a pivot table and merge it with some descriptive statistic to spot differences between the different types of cardiomyopathy. 
```{r summary, warning=FALSE}
setwd('MAGNET_GX_Results_2022')

# compute descriptive statistics for the study population and write to results directory
stargazer(sampleInfo, type = "text", title="Table 1: Summary Statistics", out="table1_summary_statistics.txt")
```
**Table 1** shows that the age in our study ranges from 15kg to 83kg and the mean weight of the participants was 81kg. We can immediately see that the minimum and maximum weights are outliers and are worth revisiting to make sure that the data is reliable. For now, we will assume that the etiology and the underlying transcriptomics measurements are correct and that the outliers for height and weight are simply mislabeled. We may decide to exclude these datapoints from our model if they are wrong or give extra consideration to them later on if they happen to reflect true values.

```{r etiology, warning=FALSE}
setwd('MAGNET_GX_Results_2022')

# create a pivot table of gender against etiology
gender <- as.data.frame(do.call(rbind, tapply(sampleInfo$gender, sampleInfo$etiology, table)))

# Extract the mean value per etiology from the summary statistics
age <- as.data.frame(do.call(rbind, tapply(sampleInfo$age, sampleInfo$etiology, summary)))['Mean']
age <- round(age, 2)
colnames(age) <- c("Avg. Age")

# the tissue source is an interesting field to check the quality of the dataset
source <- as.data.frame(do.call(rbind, tapply(sampleInfo$tissue_source, sampleInfo$etiology, table)))
colnames(source) <- c("Cardiectomy", "Donor")

# export table to png using gridExtra
png("table2_summary_etiology.png", width = 480, height = 240, units = "px")
grid.arrange(top = "Table 2: Summary Etiology", tableGrob(cbind(age, source, gender)))
dev.off()
```
**Table 2** contains summary statistics and pivot tables for patient characteristics per etiology. There is one case of HCM where the the tissue source is labeled as a donor heart.

Next, we visualize the **age and gender distribution** in different types of cardiomyopathy and compare the four etiologies.
```{r age}
age <- ggplot(sampleInfo, aes(x=age, fill=etiology)) +
  geom_histogram(position="stack", binwidth=5) +
  labs(
    title="Age distribution and etiology in cardiomyopathy",
    subtitle="study population",
    tag="Figure 1",
    x = "age",
    y = "paticipants"
  )

age + theme_bw()
```

From **Figure 1** we can see that PPCM is prevalent in women at the age of 25 to 50. The mass of the age distribution for the other etiologies lies around 55.

```{r gender}
gender <- ggplot(sampleInfo, aes(x=etiology, fill=gender)) +
	geom_bar(position="dodge") + 
	labs(
    title = "Gender and etiology in cardiomyopathy",
    subtitle = "study population",
    tag = "Figure 2",
    x = "etiology",
    y = "participants",
	)

gender + theme_bw()
```

**Figure 2** confirms that PPCM only occurs in patients with female gender. In total we have less cases of PPCM included than of other types of cardiomyopathy. In this dataset more men are diagnosed with dilated cardiomyopathy (DCM) than women. This imbalance can also be found in the general population and is most likely not a sampling error.

## Diagnostic Plots
After analyzing the sample information we examine the transcriptome of the cardiomyocytes in the different conditions and the control. We can **check the range of gene expression** to make sure that all groups follow the same distribution.

```{r expression, warning=FALSE}
gxEtiology <- gxData
# identify samples by their etiology
colnames(gxEtiology) <- sampleInfo$etiology

# all measurements
df <- melt(gxEtiology)

# etiology and gene expression
df$variable <- factor(df$variable)

# check the distribution of gene expression for all etiologies
expression <- ggplot(df, aes(x=value, color=variable)) +
  geom_density() +
  labs(
    title = "Distribution of gene expression in cardiomyopathy",
    subtitle = "mRNA expression",
    tag = "Figure 3",
    x = "values (log2 CPM)",
    color = "etiology",
    y = "genes (%/log2 CPM)",
	)

expression + theme_bw()
```

**Figure 3** shows a biomodal distribution of the original data with genes that are turned up- and down. The data is log-transformed mRNA expression counts. Hence, values below 5 might be noise.

```{r samples, warning=FALSE}
gxEtiology <- gxData

colnames(gxEtiology) <- sampleInfo$etiology
df <- cbind(melt(gxEtiology), melt(gxData)['variable'])
colnames(df) <- c("etiology","value","sample")
df$etiology <- gsub("\\..*", "", df$etiology)

samples <- ggplot(df, aes(x=sample, y=value)) +
  geom_boxplot() + 
  facet_grid(rows=vars(etiology)) +
  labs(
    title = "Distribution of gene expression per sample",
    subtitle = "mRNA expression",
    tag = "Figure 4",
    x = "samples in the dataset",
    y = "values (log2 CPM)"
	)

samples + theme_bw()
```

**Figure 4** shows where the samples appear in the dataset and the quantiles of the expression of their genes in a boxplot. If the sample contains outlier genes it is marked with a dot on the whisker. There is a batch of NF samples attached to the beginning of the dataset. It might be that those have been processed at a different time and have been added to the dataset afterwards.

## Principal Component Analysis

Next, we will **use PCA to find dimensions of the data** that can explain the etiology of cardiomyopathy. 

```{r pca}
# compute PCA decomposition using SVD
gxData.pca <- pca(t(gxData), method="svd", nPcs=5)

scores <- scores(gxData.pca)
explained_var <- sDev(gxData.pca)^2 / sum(sDev(gxData.pca)^2)
coef1 <- gxData.pca@loadings[,1]
```

```{r score}
df <- cbind(sampleInfo, scores)

PCAscore <- ggplot(df, aes(x=PC1, y=PC2, color=etiology)) +
  geom_point() +
  labs(
    title = "PCA score plot clustered by etiology",
    subtitle = "mRNA expression data",
    tag = "Figure 5",
    x = sprintf("PC1 (%1.2f%%)", 100*explained_var['PC1']),
    y = sprintf("PC2 (%1.2f%%)", 100*explained_var['PC2'])
	)

PCAscore + theme_bw()
```

**Figure 5:** The first principal component (PC1) captures 57% the variance in the dataset, but it does not have an influence on the etiology of cardiomyopathy. This might be an indication that the primary variation in the dataset is due to a batch effect or an unconsidered variables. We see that the second principal component seems to describe the variance between the non-failing heart and the three types of cardiomyopathy (15.38%).

```{r scree}
df <- as.data.frame(explained_var)
df$pc <- row.names(df)

scree <- ggplot(df) +
  geom_bar(aes(x=pc, y=explained_var), stat="identity", fill="steelblue") +
  geom_point(aes(x=pc, y=cumsum(explained_var))) + 
  geom_line(aes(x=pc, y=cumsum(explained_var)), group=1) +
  labs(
    title = "PCA scree plot of explained variance",
    tag="Figure 6",
    y = "variance explained",
    x = "principal components"
  )

scree + theme_bw()
```

## Statistical Analysis

We can use *'limma':Linear models for microarray and RNA-Seq data* to perform a differential gene expression analysis (DGEA) in cardiomyopathy. The function *lmFit* fits a linear mixed model to the gene expression. The coefficients and p-values from the different case-control contrasts can be extracted from the model and used to determine if genes are significantly differently expressed between the cases and the controls. 

In the MAGnet dataset, we have categorical variables (Etiology, Gender, Diabetes, Hypertension) that need to be included in the regression model. To accomodate for this, we can create so called 'dummy' variables that codes for the individual categories with 1's and 0's. Then, we can split up the effect of a categorical variable into individual terms.
$$
\text{Gene Expression} = \begin{cases} 
\beta_0 + \beta_1 & \text{if person is male} \\
\beta_0 & \text{if person is female}
\end{cases}
$$
For example, to describe the effect of systolic dysfunction (DCM) on gene expression, $\beta_1$ is the effect of DCM, versus the baseline model which is the not-failing heart (NF).
```{r design}
# if the baseline level is the healthy group limma will create the design matrix correctly
sampleInfo$etiology <- factor(sampleInfo$etiology, levels = c("NF", "DCM", "HCM", "PPCM"))

# creates dummy variables for categorical variables
design <- model.matrix(~0+etiology + gender + age, data=sampleInfo)
```

We need to correct for potential covariates by adding them to the model matrix above. Variables need to be added as **covariates if they explain variation in the underlying gene expression without being correlated with the etiology contrast**. This way we factor out effects on gene expression that are not based on differences in etiology.
```{r cor}
meta <- sampleInfo

levels(meta$etiology)[levels(meta$etiology)=="DCM"] <-"CM"
levels(meta$etiology)[levels(meta$etiology)=="HCM"] <-"CM"
levels(meta$etiology)[levels(meta$etiology)=="PPCM"] <-"CM"

cor(sampleInfo$age, scores)
biserial.cor(sampleInfo$age, meta$etiology)
```
Age has a slight correlation with gene expression that is not explained by the etiology present in PC5 and perhaps in PC2. Hence, I will add it as a confounding variable in the linear mixed model.

```{r cramerv}
cramerV(table(sampleInfo$Hypertension, sampleInfo$etiology))
cramerV(table(sampleInfo$Diabetes, sampleInfo$etiology))
cramerV(table(sampleInfo$tissue_source, sampleInfo$etiology))
cramerV(table(sampleInfo$race, sampleInfo$etiology))
cramerV(table(sampleInfo$gender, sampleInfo$etiology))
cramerV(table(sampleInfo$afib, sampleInfo$etiology))
```
Hypertension, tissue source, race and atrial fibrilation all have a correlation with the etiology, without having a strong effect on the gene expression as represented by the principal components. Therefore, I will include these effects in the contrasts for etiology and not correct for them.

```{r biserial}
doutlier <- which(is.na(sampleInfo$Diabetes))
biserial.cor(scores[,1][-doutlier], sampleInfo$Diabetes[-doutlier])
biserial.cor(scores[,2][-doutlier], sampleInfo$Diabetes[-doutlier])
biserial.cor(scores[,3][-doutlier], sampleInfo$Diabetes[-doutlier])
biserial.cor(scores[,4][-doutlier], sampleInfo$Diabetes[-doutlier])
biserial.cor(scores[,5][-doutlier], sampleInfo$Diabetes[-doutlier])
```
After checking the biserial correlation between the PCA scores and Diabetes, none of the principal components corresponds to the difference between diabetic and non-diabetic patients. For now, I will not explicitly correct for the effect of diabetes on cardiomyopathy.

When taking the possible co-variates into account, we might want construct a model that looks like
$$
\text{Gene Expression} = 
\begin{cases}
\beta_0 + \beta_1 & \text{if case is DCM} \\
\beta_0 & \text{if case is NF}
\end{cases} + 
\begin{cases}
\beta_2 + \beta_3 & \text{if person is male} \\
\beta_2 & \text{if person is female} 
\end{cases} + 
\beta_4 \text{ Age}
$$
```{r contrasts}
# use contrasts to create multiple codings that compare all etiologies in the same linear mixed model
contrasts <- makeContrasts(contrasts=c("etiologyDCM-etiologyNF", "etiologyHCM-etiologyNF", "etiologyPPCM-etiologyNF"), levels=colnames(design))
```
The model can be recoded by defining a contrast matrix to fit HCM vs. NF and PPCM vs. NF. Adding this matrix will allow us to compute coefficients for the three case-control scenarios.

```{r mixed linear model}
# fit linear regression for each gene
cardiomyopathy <- lmFit(gxData, design)

# compute coefficients for each case-control pair
cardiomyopathy  <- contrasts.fit(cardiomyopathy, contrasts)

# Benjamini-Hochberg correction (FDR)
cardiomyopathy <- eBayes(cardiomyopathy, trend=TRUE)

head(cardiomyopathy$coefficients)
```

```{r toptables}
# generate top-tables of p-values and log-fold changes
diffDCM <- topTable(cardiomyopathy, coef=1, n=Inf, adjust="fdr")
diffHCM <- topTable(cardiomyopathy, coef=2, n=Inf, adjust="fdr")
diffPPCM <- topTable(cardiomyopathy, coef=3, n=Inf, adjust="fdr")
```

Lastly, we extract the p-values and the log-fold changes for the three case-control scenarios from the model.
```{r dcm, echo=FALSE}
EnhancedVolcano(diffDCM, x="logFC", y="adj.P.Val", lab = rownames(diffDCM),
                title="Dilated Cardiomyopathy (DCM)",
                FCcutoff = 1.333, pCutoff = 10e-4,
                caption = 'FC cutoff, 1.333; p-value cutoff, 10e-4')
```

```{r hcm, echo=FALSE}
EnhancedVolcano(diffHCM, x="logFC", y="adj.P.Val", lab = rownames(diffHCM),
                title="Hypertrophic cardiomyopathy (HCM)",
                FCcutoff = 1.333, pCutoff = 10e-8,
                caption = 'FC cutoff, 1.333; p-value cutoff, 10e-4')
```

```{r ppcm, echo=FALSE}
EnhancedVolcano(diffPPCM, x="logFC", y="adj.P.Val", lab = rownames(diffPPCM),
                title="Peripartum cardiomyopathy (PPCM)",
                FCcutoff = 1.333, pCutoff = 10e-4,
                caption = 'FC cutoff, 1.333; p-value cutoff, 10e-4')
```

The above volcano plots can be used as a diagnostic tool to check the results of the differential gene expression analysis.

## Additional Gene Annotation

Afterwards, we add additional gene annotations from the [Ensembl](http://www.ensembl.org/index.html) database to our differential gene expression dataset. We add the common gene symbol, a short description and the chromosome on which the gene is located to the final annotated dataset.
```{r ensemble, warning=FALSE}
# connect to the Ensembl genes database and select species
ensembl = useEnsembl(biomart="genes", dataset="hsapiens_gene_ensembl")

annotations <- getBM(attributes=c('ensembl_gene_id', 'hgnc_symbol', 'description', 'chromosome_name'), 
    filters = 'ensembl_gene_id', values = rownames(gxData), mart = ensembl)

colnames(diffDCM) <- paste("DCMvsNF_",colnames(diffDCM))
colnames(diffHCM) <- paste("HCMvsNF_",colnames(diffHCM))
colnames(diffPPCM) <- paste("PPCMvsNF_",colnames(diffPPCM))

annotations <- merge(annotations, diffDCM, by.x="ensembl_gene_id", by.y="row.names")
annotations <- merge(annotations, diffHCM, by.x="ensembl_gene_id", by.y="row.names")
annotations <- merge(annotations, diffPPCM, by.x="ensembl_gene_id", by.y="row.names")

str(annotations)
```

### Relative Expression Levels

As a last step, we convert the log-2 transformed CPM values to TPM values by dividing them with the exon length for each gene. Then, we can **compare the relative expression of each gene** to i.e. set universal noise thresholds for all genes or to visualize their relative expression in an enriched pathway.
$$
TPM=\frac{\text{total reads mapped to gene} * 10^3}{\text{gene length in bp}} 
$$
```{r convert}
cpm2tpm <- function(x) {
	.t <- 2^(x) * 1E3 / exonLength[, 1]
}
gxData.tpm <- cpm2tpm(gxData)
```

We define the average expression of the genes in the **Y chromosomes** for female subjects as the **noise threshold**. For each gene a field is added that indicates whether it is on average expressed above noise level in our analysis.
```{r noise, warning=FALSE}
gxGender <- gxData.tpm
colnames(gxGender) <- sampleInfo$gender

# pick the ensembl gene id of genes in the Y chromosome
Y <- annotations[annotations$chromosome_name == "Y",]$ensembl_gene_id
df <- melt(gxGender[Y,])

# average expression of Y chromosome genes in females
noise_level <- mean(df[df$variable == "Female",]$value)

# define all genes with avg expression below that as noise
noise <- as.data.frame(rowMeans(gxData.tpm) < noise_level)
colnames(noise) <- c("noise")

annotations <- merge(annotations, noise, by.x="ensembl_gene_id", by.y="row.names")
```

We can calculate the average expression per etiology for later visualizing it in an enriched pathway.
```{r avgs}
gxData.avg <- gxData.tpm
colnames(gxData.avg) <- sampleInfo$etiology

# aggregate average expression by column name from the data converted to TPM values
gxData.avg <- t(rowsum(t(gxData.avg), names(gxData.avg))/c(table(names(gxData.avg))))

colnames(gxData.avg) <- paste("Avg_TPM_", colnames(gxData.avg))
annotations <- merge(annotations, gxData.avg, by.x="ensembl_gene_id", by.y="row.names")
```

The following code shows the five highest expressed genes across the four etiologies, just as an interesting side note.
```{r select}
ordered_avg_expr_tpm <- gxData.tpm[order(rowMeans(gxData.tpm)),]

highest_tpm <- tail(ordered_avg_expr_tpm, n=5)
highest_genes <- row.names(highest_tpm)
```

We can present them in a heatmap with a dendrogram to cluster genes based on their expression.
```{r genes}
selectedGenes <- c(highest_genes)
pheatmap(gxData.avg[selectedGenes,], scale = 'row', treeheight_col = 0, show_rownames = TRUE)
```

The five genes that are the most expressed in the non-failing heart are all related to OXPHOS:
```{r genecard, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
highest <- "
| Symbol         | Description       | Category  |
|:---------------|:----------------:|:-----------------|
| MT-ND4       | Mitochondrially Encoded NADH:Ubiquinone Oxidoreductase Core Subunit 4 | Protein Coding|
| MT-CO2       | Mitochondrially Encoded Cytochrome C Oxidase II | Protein Coding|
| MT-CO3       | Mitochondrially Encoded Cytochrome C Oxidase III | Protein Coding|
| MT-ATP6       | Mitochondrially Encoded ATP Synthase Membrane Subunit 6| Protein Coding|
| MT-CO1       | Mitochondrially Encoded Cytochrome C Oxidase I | Protein Coding|
"
cat(highest)
```

### Exporting Results

Prepare the results for further visualization or pathway analysis by writing them to a tab-delimited text file. The exported file can be sorted by the adjusted p-value for each contrast. Then, the top N genes can be read from the list.
```{r export, warning=FALSE, results='hide'}
# write data frame with gene annotations and differential gene expression to a tab-delimited file
write.table(annotations, file="MAGNET_DifferentialGeneExpression_annotated.txt", sep="\t", row.names=FALSE)

# uncomment to collect figures
setwd('MAGNET_GX_Results_2022')

# exploratory data analysis
png("figure_1_age_distribution_etiology.png")
age + theme_bw()
dev.off()

png("figure_2_gender_and_etiology.png")
gender + theme_bw()
dev.off()

# diagnostic plots
png("figure_3_distribution_gene_expression.png")
expression + theme_bw()
dev.off()

png("figure_4_distribution_samples.png")
samples + theme_bw()
dev.off()

# PCA
png("figure_5_pca_scores_plot.png")
PCAscore + theme_bw()
dev.off()

png("figure_6_pca_scree_plot.png")
scree + theme_bw()
dev.off()
```

After running the whole script all figures will be available in the *MAGNET_GX_Results_2022* directory. I decided that I'd like to export the dataset in one single file, such that I can sort columns, apply filters and extract gene lists on the go by loading the file completely in Excel for instance. The annotated DGEA dataset is written as a text file *MAGNET_DifferentialGeneExpression_annotated.txt* directly to the current working directory. 